AR-js-org/AR.js
# Partial fix of #110 (distant augmented data not visible even if within the viewing frustum)

**⚠️ All PRs have to be done versus 'dev' branch, so be aware of that, or we'll close your issue ⚠️**

**What kind of change does this PR introduce?**

Partial fix of issue #110. Allows AR content distant from the current position to be viewed, up to the far clip plane.

**Can it be referenced to an Issue? If so what is the issue # ?**
#110 

**How can we test it?**


Use test application at
https://hikar.org/webapp/

On a mobile device, will use GPS location. Or, on a desktop, use hardcoded location:
https://hikar.org/webapp/?lat=51.049&lon=-0.723

Ensure you hard-refresh your browser, to make sure you're not using an earlier version without this feature.
**Summary**

Camera feed is rendered using a WebGL texture, making use of `THREE.VideoTexture` within an A-Frame component.

**Does this PR introduce a breaking change?**
No

**Please TEST your PR before proposing it. Specify here what device you have used for tests, version of OS and version of Browser**
Laptop using Ubuntu 18.04 and latest versions of Firefox and Chrome; Pixel 3 running Android 10 and again latest versions of Firefox and Chrome.

**Other information**
Some detail about this along with the current shortcomings and flaws. I'd like to invite discussion on whether this PR fits in with the overall AR.js architecture.

First of all I have created a new A-Frame component, `arjs-webcam-texture`, within the aframe/src/location-based directory. This will use `THREE.VideoTexture` to stream the webcam video to a texture.

I've had to do a bit of a hack to do this; I first attempted it using pure three.js and then within the `arjs-webcam-texture` A-Frame component. Namely, a second camera, an `OrthographicCamera` (as we don't want perspective for the texture) is created, and then a geometry is created to fill the whole screen.
A mesh is created using this geometry and the texture, and attached to a *separate* scene. (I'm not sure whether you can have one scene with two entirely separate cameras for completely different purposes).

Then in the component's `tick()`, this separate scene is rendered using the orthographic camera. This appears to be the recommended way of doing this in plain three.js, and it appears to work in A-Frame also. Some parameters need to be set to prevent the screen clearing between rendering the two scenes.

The other change is that in the `arjs-system`, a new property `videoTexture` is added. If this is true, then the new `arjs-webcam-texture` component is used (by dynamically creating a new `a-entity` using it and attaching it to the scene), and the existing `ARjs.Session` is not used at all.

This basically works, but with one main flaw, the field of view of the AR view does not match that of the hardware camera device. Actually, looking quickly through the AR.js source code, there does not appear to be anything there yet which obtains the FOV of the hardware camera device, is this correct?

I believe this is an issue with the augmented content, not the live camera stream, as the FOV of the live camera stream appears to match that of the native camera app of the mobile device (Pixel 3) used to test.

Would like to get feedback on this PR and whether my proposed approach fits in with the long-term plans of AR.js and in particular, how compatible this is with the NFT feature and any future image detection functionality.
